toolchain/check/pattern_match.cpp

```diff
@@ -55,22 +56,57 @@ class MatchContext {
   }
 
   // Adds a work item to the stack. Cannot be called after Finish().
-  auto AddWork(WorkItem work_item) -> void {
-    CARBON_CHECK(!result_.is_valid());
-    stack_.push_back(work_item);
-  }
+  auto AddWork(WorkItem work_item) -> void { stack_.push_back(work_item); }
 
   // Returns the next work item to process.
   auto NextWorkItem() -> WorkItem { return stack_.pop_back_val(); }
 
-  // Sets the result of this pattern matching operation. Must not be called
-  // when there is still pending work, except to report an error.
+  // Allocates the next unallocated RuntimeParamIndex, starting from 0.
+  auto NextRuntimeIndex() -> SemIR::RuntimeParamIndex {
+    auto result = next_index_;
+    ++next_index_.index;
+    return result;
+  }
+
+  // TODO: Eliminate the caller/callee API split below, by restructuring
+  // CallerPatternMatch to operate on the whole pattern.
+
+  // Sets the result of this pattern matching operation. Must not be called when
+  // there is still pending work, except to report an error, or called more than
+  // once between calls to ConsumeResult. Valid only during caller matching.
   auto Finish(SemIR::InstId result) -> void {
     CARBON_CHECK(!HasWork() || result == SemIR::InstId::BuiltinError);
+    CARBON_CHECK(kind_ == MatchKind::Caller);
+    CARBON_CHECK(result_ == SemIR::InstId::Invalid);
     result_ = result;
   }
 
-  auto result() const -> SemIR::InstId { return result_; }
+  // Consumes and returns the result stored by Finish. Valid only during caller
+  // matching.
+  auto ConsumeResult() -> SemIR::InstId {
+    CARBON_CHECK(stack_.empty() || result_ == SemIR::InstId::BuiltinError);
+    CARBON_CHECK(kind_ == MatchKind::Caller);
+    return std::exchange(result_, SemIR::InstId::Invalid);
+  }
+
+  // Records that `bind_name_id` is the ID of an "inner parameter", i.e. an inst
+  // in the decl block that exposes a parameter to the function body. Valid
+  // only during callee pattern matching.
+  auto RecordInnerParam(SemIR::InstId bind_name_id) {
+    CARBON_CHECK(kind_ == MatchKind::Callee);
+    inner_param_insts_.push_back(bind_name_id);
+  }
+
+  // Allocates an InstBlock containing the inner parameters recorded by
+  // RecordInnerParam since the last call to this function (if any), and returns
+  // its ID. Valid only during callee pattern matching.
+  auto ConsumeInnerParams(Context& context) -> SemIR::InstBlockId {
+    CARBON_CHECK(stack_.empty());
+    CARBON_CHECK(kind_ == MatchKind::Callee);
+    auto block_id = context.inst_blocks().Add(inner_param_insts_);
+    inner_param_insts_.clear();
+    return block_id;
+  }
```

Comment:
If I'm reading the code below correctly, what you're calling "inner parameters" here are the bindings in the pattern. Is that right? If so, would it make sense to use that terminology here, or are you planning for this to change to instead be a list of parameters? (Either way, I think it'd be good for the comment on `RecordInnerParam` to clarify, perhaps with an example.)

Reply:
Yeah, I think I was way overthinking this; in practice an inner parameter is always a name binding and vice versa, so there's no need for a separate term. Is this clear enough now?

Reply:
Yeah, that works for me. Thanks!